libbpf-tools/filetop.h

```diff
@@ -0,0 +1,31 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+#ifndef __FILETOP_H
+#define __FILETOP_H
+
+#define PATH_MAX	4096
+#define TASK_COMM_LEN	16
+
+enum op {
+	READ,
+	WRITE,
+};
+
+struct file_id {
```

Comment:
so the key -> val mapping of the python program was:
`(pid, name_len, comm, name, type) -> (reads, writes, rbytes, wbytes)`

while this tool's is:
`(file, pid, tid) -> (reads, rbytes, writes, wbytes, pid, tid, filename, comm, type)`
where file is a `struct file*`

I think the new way is better but wanted to run through some edge cases and ask questions to make sure I'm understanding correctly:
* If a file is closed and re-opened, do we expect to see a different `struct file *`? Similarly, is there any chance we get the same `struct file *` if we close file A and open file B shortly after?
  * Latter hypothetical is more concerning to me as we'd be attributing io to wrong filename
* Why do `pid, tid` exist in both key and value?
* Was the purpose of this change to avoid reading `comm` and `filename` unnecessarily?

Reply:
> * If a file is closed and re-opened, do we expect to see a different struct file *? Similarly, is there any chance we get the same struct file * if we close file A and open file B shortly after?
>     *  Latter hypothetical is more concerning to me as we'd be attributing io to wrong filename

I can't answer the first question. :(

But since the original tool only use the filename (without full file path), I think that is more unreliable.
Say, the same process writes to `/foo/bar/xxx` and `/fuz/baz/xxx` would be added up to the same key.

> * Why do pid, tid exist in both key and value?

pid and tid exist in value for userspace output, since we sort and output the values only.

> * Was the purpose of this change to avoid reading comm and filename unnecessarily?

Yes, I think this is better from the perspective of performance.

Reply:
> But since the original tool only use the filename (without full file path), I think that is more unreliable.
Say, the same process writes to /foo/bar/xxx and /fuz/baz/xxx would be added up to the same key.

Ah, thanks for noting that, that's definitely worth improving on.

Unfortunately after digging into where `struct file *`s are coming from I think the scenarios I was worried about can happen. When the kernel [allocates a new `struct file`](https://elixir.bootlin.com/linux/latest/source/fs/file_table.c#L96) it uses memory provided by a [slab allocator](https://elixir.bootlin.com/linux/latest/source/fs/file_table.c#L41) that just serves `struct file`-sized memory. 

So for my two hypotheticals: if a file is closed and re-opened, there's no guarantee of seeing the same `struct file *` or a different one. Similarly for the more concerning one, two different files could have the same `struct file *` if first one is closed before second opened.

Not sure how often this happens in practice, but given that this tool is more likely to be used on systems that are doing a lot of "file stuff" and therefore more likely to trigger these cases, we should find some other identifier.  

The inode's i_ino and device number of the superblock (`inode->i_sb->s_dev`) should do it.

Reply:
Thanks.

> The inode's i_ino and device number of the superblock (inode->i_sb->s_dev) should do it.

Will do.

Reply:
I read this [page](https://www.oreilly.com/library/view/linux-device-drivers/0596000081/ch03s02.html) which says:

> Every time the kernel calls a device driver, it tells the driver which device is being acted upon. The major and minor numbers are paired in a single data type that the driver uses to identify a particular device. The combined device number (the major and minor numbers concatenated together) resides in the field i_rdev of the inode structure, which we introduce later. Some driver functions receive a pointer to struct inode as the first argument. So if you call the pointer inode (as most driver writers do), the function can extract the device number by looking at inode->i_rdev.

Also, the kernel has two functions `imajor` and `iminor` which extract device number from `inode->i_rdev`.

https://elixir.bootlin.com/linux/latest/source/include/linux/fs.h#L875

I think it's reasonable to just use `inode->i_rdev`, WDYT?

Reply:
Updated. :)

Reply:
I think just using the `rdev` directly is fine, no need to pull out major/minor